### PR TITLE
[autograd] match 0-dim gradients device type regardless of subclassness

### PR DESCRIPTION
**Pull request by xmfan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #160165

Not sure if there some subclasses where the outer.dim() == 0 but you wouldn't want to move it?

FIXES https://github.com/pytorch/pytorch/issues/160084

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -979,13 +979,13 @@ static void validate_outputs_impl(
...
-      // should be eventually removed
-      if (!(metadata.is_tensor_subclass() ||
-            grad.unsafeGetTensorImpl()->is_python_dispatch())) {
-        if (grad.dim() == 0) {
-          grad = grad.to(metadata.device());
-        } else {
+      if (grad.dim() == 0) {
+        grad = grad.to(metadata.device());
+      } else {
+        // quick hack for: https://github.com/pytorch/pytorch/issues/65016 but
```
Comment (albanD) [inline torch/csrc/autograd/engine.cpp:985]: @ezyang now that we do have proper wrapper subclass, should we remove this as we don't have to use meta device for all subclasses anymore

File: test/dynamo/test_repros.py [hunk 2]
```diff
@@ -7673,6 +7673,32 @@ def forward(self, x):
         out2 = torch.compile(model, backend="eager")(input.clone())
         self.assertEqual(out1, out2)
 
+    @requires_cuda
```
Comment (albanD) [inline test/dynamo/test_repros.py:7676]: Please add test for this in test_autograd.py via generic device tests to ensure this works fine on all device.

**Review by albanD — COMMENTED** [review]
Sounds good. A bit more testing is needed.

File: test/test_autograd.py [hunk 3]
```diff
@@ -12396,6 +12396,28 @@ def test_resize_version_bump(self, device):
...
+        class RegressionModel(torch.nn.Module):
+            def __init__(self, a=0, b=0):
+                super().__init__()
+                self.a = torch.nn.Parameter(torch.tensor(a).float())
+                self.b = torch.nn.Parameter(torch.tensor(b).float())
+
+            def forward(self, x):
+                return x * self.a + self.b
+
+        model = RegressionModel()
```
Comment (albanD) [inline test/test_autograd.py]:
```suggestion

        # Keep the model on cpu as we do want to test the mixed cpu/accelerator behavior here
        model = RegressionModel()
```

**Review by albanD — APPROVED** [review]
Small nit, SGTM otherwise

File: test/test_python_dispatch.py [hunk 4]
```diff
@@ -1718,49 +1717,6 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
...
 
-    def test_subclass_autograd_device_check(self) -> None:
-        class NonWrapperSubclass(torch.Tensor):
-            elem: torch.Tensor
-
-            __slots__ = ["elem"]
-
-            @staticmethod
-            def __new__(cls, elem, *args, **kwargs):
-                # Wrong device here!
```
Comment (xmfan) [inline test/test_python_dispatch.py:1729]: @albanD This test intentionally diverges the outer tensor's device from the inner tensor's device, is this something we still want to support? Assuming it's not, the test wouldn't make sense anymore if we kept the two in sync so I just removed it
